source/acquaman/BioXAS/BioXASXASScanActionController.cpp

```diff
@@ -247,233 +35,5 @@ void BioXASXASScanActionController::createScanAssembler()
-		absorbanceCorrectedSource->setExpression(QString("ln(%1/%2)").arg(i1CorrectedDetectorSource->name(), i2CorrectedDetectorSource->name()));
-
-		scan_->addAnalyzedDataSource(absorbanceCorrectedSource, true, false);
-	}
-
-	// Create analyzed data source for the derivative of the absorbance, dark current corrected values.
-
-	AM1DDerivativeAB *derivAbsorbanceCorrectedSource = 0;
-
-	if (absorbanceCorrectedSource) {
-		derivAbsorbanceCorrectedSource = new AM1DDerivativeAB("DerivAbsorbance_DarkCorrect");
-		derivAbsorbanceCorrectedSource->setInputDataSources(QList<AMDataSource*>() << absorbanceCorrectedSource);
-
-		scan_->addAnalyzedDataSource(derivAbsorbanceCorrectedSource, true, false);
-	}
-
-	// Create analyzed data source for the Ge 32-el detector.
-
-	AMXRFDetector *ge32Detector = BioXASBeamline::bioXAS()->ge32ElementDetector();
-
-	if (ge32Detector) {
-
-		int ge32DetectorIndex = scan_->indexOfDataSource(ge32Detector->name());
-
-		if (ge32DetectorIndex != -1) {
-
-			if (zebraTriggerSource) {
-				zebraTriggerSource->addDetector(ge32Detector);
-				zebraTriggerSource->addDetectorManager(ge32Detector);
-			}
-
-			// Clear any previous regions.
-
-			ge32Detector->removeAllRegionsOfInterest();
-
-			// Iterate through each region in the configuration.
-			// Create analysis block for each region, add ge32Detector spectra source as input source for each.
-			// Add analysis block to the scan and to the ge32Detector.
-			// Create normalized analysis block for each region, add to scan.
-
-			AMDataSource *spectraSource = scan_->dataSourceAt(ge32DetectorIndex);
-			QString edgeSymbol = bioXASConfiguration_->edge().split(" ").first();
-			bool canNormalize = (i0DetectorSource || i0CorrectedDetectorSource);
-
-			foreach (AMRegionOfInterest *region, bioXASConfiguration_->regionsOfInterest()){
-
-				AMRegionOfInterestAB *regionAB = (AMRegionOfInterestAB *)region->valueSource();
-
-				AMRegionOfInterestAB *newRegion = new AMRegionOfInterestAB(regionAB->name().remove(' '));
-				newRegion->setBinningRange(regionAB->binningRange());
-				newRegion->setInputDataSources(QList<AMDataSource *>() << spectraSource);
-
-				scan_->addAnalyzedDataSource(newRegion, false, true);
-				ge32Detector->addRegionOfInterest(region);
-
-				if (canNormalize) {
-					AMDataSource *normalizationSource = (i0CorrectedDetectorSource != 0) ? i0CorrectedDetectorSource : i0DetectorSource;
-
-					AM1DNormalizationAB *normalizedRegion = new AM1DNormalizationAB(QString("norm_%1").arg(newRegion->name()));
-					normalizedRegion->setInputDataSources(QList<AMDataSource *>() << newRegion << normalizationSource);
-					normalizedRegion->setDataName(newRegion->name());
-					normalizedRegion->setNormalizationName(normalizationSource->name());
-
-					scan_->addAnalyzedDataSource(normalizedRegion, newRegion->name().contains(edgeSymbol), !newRegion->name().contains(edgeSymbol));
-				}
-			}
-		}
-	}
+	BioXASBeamline::bioXAS()->buildScan(scan_);
```

Comment:
This part I agree with @iainworkman where the different scan types can become an issue.